src/detect-mqtt-connack-sessionpresent.c

```diff
@@ -0,0 +1,328 @@
+    sigmatch_table[DETECT_AL_MQTT_CONNACK_SESSION_PRESENT].url = DOC_URL DOC_VERSION "/rules/mqtt-keywords.html#mqtt-connack-session-present";
+    sigmatch_table[DETECT_AL_MQTT_CONNACK_SESSION_PRESENT].AppLayerTxMatch = DetectMQTTConnackSessionPresentMatch;
+    sigmatch_table[DETECT_AL_MQTT_CONNACK_SESSION_PRESENT].Setup = DetectMQTTConnackSessionPresentSetup;
+    sigmatch_table[DETECT_AL_MQTT_CONNACK_SESSION_PRESENT].Free  = DetectMQTTConnackSessionPresentFree;
+    sigmatch_table[DETECT_AL_MQTT_CONNACK_SESSION_PRESENT].RegisterTests = MQTTConnackSessionPresentRegisterTests;
+
+    DetectSetupParseRegexes(PARSE_REGEX, &parse_regex);
+
+    DetectAppLayerInspectEngineRegister("mqtt.connack.session_present",
+            ALPROTO_MQTT, SIG_FLAG_TOSERVER, 1,
+            DetectEngineInspectMQTTConnackSessionPresentGeneric);
+
+    mqtt_connack_session_present_id = DetectBufferTypeGetByName("mqtt.connack.session_present");
+}
+
+static int DetectEngineInspectMQTTConnackSessionPresentGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id)
+{
+    return DetectEngineInspectGenericList(tv, de_ctx, det_ctx, s, smd,
+                                          f, flags, alstate, txv, tx_id);
+}
+
+/**
+ * \internal
+ * \brief Function to match session_present flag of an MQTT CONNACK message
+ *
+ * \param det_ctx Pointer to the pattern matcher thread.
+ * \param f       Pointer to the current flow.
+ * \param flags   Flags.
+ * \param state   App layer state.
+ * \param txv     Pointer to the transaction.
+ * \param s       Pointer to the Signature.
+ * \param ctx     Pointer to the sigmatch that we will cast into DetectMQTTConnackSessionPresentData.
+ *
+ * \retval 0 no match.
+ * \retval 1 match.
+ */
+static int DetectMQTTConnackSessionPresentMatch(DetectEngineThreadCtx *det_ctx,
+                               Flow *f, uint8_t flags, void *state,
+                               void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+{
+    const DetectMQTTConnackSessionPresentData *de = (const DetectMQTTConnackSessionPresentData *)ctx;
+    bool value = false;
+
+    if (!de)
+        return 0;
+
+    if (rs_mqtt_tx_get_connack_sessionpresent(txv, &value) ==0 ) {
+        return 0;
+    }
+    if (value != de->session_present) {
+        return 0;
+    }
+
+    return 1;
+}
+
+/**
+ * \internal
+ * \brief This function is used to parse options passed via mqtt.connack.session_present: keyword
+ *
+ * \param rawstr Pointer to the user provided options
+ *
+ * \retval de pointer to DetectMQTTConnackSessionPresentData on success
+ * \retval NULL on failure
+ */
+static DetectMQTTConnackSessionPresentData *DetectMQTTConnackSessionPresentParse(const char *rawstr)
+{
+    DetectMQTTConnackSessionPresentData *de = NULL;
+#define MAX_SUBSTRINGS 30
```

Comment:
Idea : Could we rather do this parsing in Rust ?

Reply:
See below.